File: tools/library-analyzers/cpp/tests/symbols.cpp
// Integration test for the ce-cpp AST symbol analyzer (spec §§6.7, 12.5,
// 13.1; plan 047).
//
// The test loads the checked-in `cpp-symbols-request.json` fixture under the
// fixture tree, runs `analyzeSymbolsForTarget` against each library, folds
// the resulting `SymbolAnalysis` values into a synthetic `AnalysisResponse`,
// and compares the JSON serialization byte-for-byte with
// `cpp-symbols-response.json`. Set `UPDATE_EXPECT=1` to rewrite the
// checked-in fixture from a green run.
//
// In addition to the fixture-driven check, a handful of targeted assertions
// verify kind classification, overload handling, and location coordinates
// so a semantic drift produces a precise error before the byte diff runs.

#include "compile_profile.hpp"
#include "protocol.hpp"
#include "symbols.hpp"

#include <algorithm>
#include <cstdlib>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

namespace fs = std::filesystem;

namespace {

int failures = 0;

#define CE_EXPECT_TRUE(expr)                                                                    \
    do {                                                                                        \
        if (!(expr)) {                                                                          \
            std::cerr << __FILE__ << ":" << __LINE__ << ": expectation failed: " << #expr       \
                      << std::endl;                                                             \
            ++failures;                                                                         \
        }                                                                                       \
    } while (0)

#define CE_EXPECT_EQ(a, b)                                                                    \
    do {                                                                                      \
        auto _a = (a);                                                                        \
        auto _b = (b);                                                                        \
        if (!(_a == _b)) {                                                                    \
            std::cerr << __FILE__ << ":" << __LINE__ << ": expected equality " << #a << " == "\
                      << #b << std::endl;                                                     \
            ++failures;                                                                       \
        }                                                                                     \
    } while (0)

std::string read_file(const fs::path& path) {
    std::ifstream f(path);
    if (!f.is_open()) {
        std::cerr << "failed to open fixture: " << path << std::endl;
        std::abort();
    }
    std::ostringstream ss;
    ss << f.rdbuf();
    return ss.str();
}

void write_file(const fs::path& path, const std::string& contents) {
    std::ofstream f(path);
    if (!f.is_open()) {
        std::cerr << "failed to open fixture for write: " << path << std::endl;
        std::abort();
    }
    f << contents;
}

std::string chomp(const std::string& s) {
    if (!s.empty() && s.back() == '\n') return s.substr(0, s.size() - 1);
    return s;
}

fs::path tree_root() { return fs::path(CE_CPP_TREE_DIR); }
fs::path protocol_fixture_dir() { return fs::path(CE_CPP_PROTOCOL_FIXTURE_DIR); }
fs::path expected_fixture_path() {
    return protocol_fixture_dir() / "cpp-symbols-response.json";
}

/// Build the in-memory `CompileProfile` the symbol tests need. Adds the
/// symbol fixture directory as an include root so `#include "basic.hpp"`
/// style headers inside the tree resolve without touching the on-disk
/// `compile-profile.toml`.
ce_cpp::CompileProfile build_test_profile() {
    ce_cpp::CompileProfile p;
    p.repository_root = fs::canonical(tree_root());
    p.cxx_standard = "c++20";
    p.include_roots.push_back(fs::canonical(tree_root() / "libraries/cpp/symbols"));
    p.include_roots.push_back(fs::canonical(tree_root() / "libraries/cpp"));
    return p;
}

ce_cpp::LibraryAnalysis run_target(const ce_cpp::CompileProfile& profile,
                                   const std::string& target_path) {
    fs::path source = tree_root() / target_path;
    auto outcome = ce_cpp::analyzeSymbolsForTarget(profile, source, target_path);
    ce_cpp::LibraryAnalysis la;
    la.path = target_path;
    la.dependency_analysis.state = ce_cpp::AnalysisState::Complete;
    la.symbol_analysis = std::move(outcome.analysis);
    return la;
}

std::string serialize_payload(const std::vector<ce_cpp::LibraryAnalysis>& libraries) {
    ce_cpp::AnalysisResponse resp;
    resp.schema_version = ce_cpp::SCHEMA_VERSION;
    resp.adapter.name = std::string(ce_cpp::ADAPTER_NAME);
    resp.adapter.version = std::string(ce_cpp::ADAPTER_VERSION);
    ce_cpp::ToolchainIdentity clang;
    clang.name = "clang";
    clang.version = "22.1.0";
    clang.target = std::string("x86_64-unknown-linux-gnu");
    resp.adapter.toolchains.push_back(std::move(clang));
    resp.libraries = libraries;
    return ce_cpp::serialize_response(resp);
}

const ce_cpp::Symbol* find_symbol(const ce_cpp::SymbolAnalysis& sa, const std::string& name,
                                  const std::string& kind) {
    for (const auto& s : sa.symbols) {
        if (s.name == name && s.kind == kind) return &s;
    }
    return nullptr;
}

size_t count_kind(const ce_cpp::SymbolAnalysis& sa, const std::string& kind) {
    size_t c = 0;
    for (const auto& s : sa.symbols) {
        if (s.kind == kind) ++c;
    }
    return c;
}

void run_semantic_checks(const std::vector<ce_cpp::LibraryAnalysis>& libraries) {
    // Locate each library analysis by its wire path.
    auto get = [&](const std::string& p) -> const ce_cpp::LibraryAnalysis& {
        for (const auto& la : libraries) {
            if (la.path == p) return la;
        }
        std::cerr << "missing library analysis: " << p << std::endl;
        std::abort();
    };

    const auto& basic = get("libraries/cpp/symbols/basic.hpp").symbol_analysis;
    CE_EXPECT_TRUE(basic.state == ce_cpp::AnalysisState::Complete ||
                   basic.state == ce_cpp::AnalysisState::Partial);
    // Namespace and enclosed class-like decls.
    CE_EXPECT_TRUE(find_symbol(basic, "algebra", "type") != nullptr);
    CE_EXPECT_TRUE(find_symbol(basic, "Point", "class") != nullptr);
    CE_EXPECT_TRUE(find_symbol(basic, "Color", "class") != nullptr);
    CE_EXPECT_TRUE(find_symbol(basic, "Bytes", "class") != nullptr);
    // Enum + enumerator.
    CE_EXPECT_TRUE(find_symbol(basic, "Signal", "type") != nullptr);
    CE_EXPECT_TRUE(find_symbol(basic, "Low", "value") != nullptr);
    CE_EXPECT_TRUE(find_symbol(basic, "High", "value") != nullptr);
    // Alias + typedef map to type.
    CE_EXPECT_TRUE(find_symbol(basic, "Coord", "type") != nullptr);
    CE_EXPECT_TRUE(find_symbol(basic, "Weight", "type") != nullptr);
    // Concept.
    CE_EXPECT_TRUE(find_symbol(basic, "Addable", "concept") != nullptr);
    // Template class.
    CE_EXPECT_TRUE(find_symbol(basic, "Optional", "class") != nullptr);
    // Template function + overloaded free functions.
    CE_EXPECT_TRUE(find_symbol(basic, "identity", "function") != nullptr);
    CE_EXPECT_EQ(count_kind(basic, "function"), size_t{3});  // identity + zero + zero
    // Constructor / destructor / method / operator all classify as method.
    CE_EXPECT_TRUE(find_symbol(basic, "Point", "method") != nullptr);
    CE_EXPECT_TRUE(find_symbol(basic, "~Point", "method") != nullptr);
    CE_EXPECT_TRUE(find_symbol(basic, "shifted", "method") != nullptr);
    CE_EXPECT_TRUE(find_symbol(basic, "origin", "method") != nullptr);
    CE_EXPECT_TRUE(find_symbol(basic, "operator+", "method") != nullptr);
    // Value: variable + enumerator.
    CE_EXPECT_TRUE(find_symbol(basic, "PI", "value") != nullptr);
    CE_EXPECT_TRUE(find_symbol(basic, "Red", "value") != nullptr);

    // Overloads: two `zero` free functions, both emitted, both with the same
    // qualified_name but different signatures.
    size_t zero_count = 0;
    for (const auto& s : basic.symbols) {
        if (s.name == "zero" && s.kind == "function") {
            ++zero_count;
            CE_EXPECT_TRUE(s.qualified_name.has_value());
            CE_EXPECT_EQ(*s.qualified_name, std::string("algebra::zero"));
            CE_EXPECT_TRUE(s.signature.has_value());
        }
    }
    CE_EXPECT_EQ(zero_count, size_t{2});

    // Nested + anonymous namespace symbol behavior.
    const auto& nested = get("libraries/cpp/symbols/nested.hpp").symbol_analysis;
    CE_EXPECT_TRUE(find_symbol(nested, "outer", "type") != nullptr);
    CE_EXPECT_TRUE(find_symbol(nested, "inner", "type") != nullptr);
    CE_EXPECT_TRUE(find_symbol(nested, "Inner", "class") != nullptr);
    if (const auto* inner_class = find_symbol(nested, "Inner", "class")) {
        CE_EXPECT_TRUE(inner_class->qualified_name.has_value());
        CE_EXPECT_EQ(*inner_class->qualified_name,
                     std::string("outer::inner::Inner"));
    }
    CE_EXPECT_TRUE(find_symbol(nested, "Hidden", "class") != nullptr);
    if (const auto* hidden = find_symbol(nested, "Hidden", "class")) {
        CE_EXPECT_TRUE(hidden->qualified_name.has_value());
        // Anonymous namespace contributes `(anonymous namespace)` to the
        // qualified name (adapter-private token, not seen by the core).
        CE_EXPECT_TRUE(hidden->qualified_name->find("(anonymous namespace)") !=
                       std::string::npos);
    }
    CE_EXPECT_TRUE(find_symbol(nested, "top", "function") != nullptr);

    // Unicode names: bytes-wise columns and Unicode identifiers pass through.
    const auto& unicode = get("libraries/cpp/symbols/unicode.hpp").symbol_analysis;
    CE_EXPECT_TRUE(find_symbol(unicode, u8"東京", "type") != nullptr);
    CE_EXPECT_TRUE(find_symbol(unicode, u8"街", "class") != nullptr);
    CE_EXPECT_TRUE(find_symbol(unicode, u8"こんにちは", "function") != nullptr);
}

// ─── Task 2: source-location and recovery semantics ────────────────────────

/// Macro spelling/expansion: declarations synthesized inside a macro body
/// have their spelling location inside the macro definition, so the
/// main-file filter suppresses them. Non-macro decls emitted alongside are
/// still returned.
void test_macro_expansions_are_filtered() {
    ce_cpp::CompileProfile profile = build_test_profile();
    auto la = run_target(profile, "libraries/cpp/symbols/macros.hpp");
    const auto& sa = la.symbol_analysis;
    CE_EXPECT_TRUE(find_symbol(sa, "Real", "class") != nullptr);
    // A `CE_DECLARE_STRUCT(FromMacro)` invocation may either project no
    // symbol at all or emit the record at the invocation site — accept
    // both, but never emit at the macro definition line.
    for (const auto& s : sa.symbols) {
        if (s.location.has_value()) {
            CE_EXPECT_TRUE(s.location->start.line > 3);
        }
    }
}

/// CRLF handling: Clang counts `\r\n` as one line terminator, so decls on
/// the physical Nth line report `line = N`, not `N + 1`.
void test_crlf_line_numbers() {
    ce_cpp::CompileProfile profile = build_test_profile();
    auto la = run_target(profile, "libraries/cpp/symbols/crlf.hpp");
    const auto& sa = la.symbol_analysis;
    const auto* first = find_symbol(sa, "First", "class");
    CE_EXPECT_TRUE(first != nullptr);
    if (first != nullptr && first->location.has_value()) {
        CE_EXPECT_EQ(first->location->start.line, uint32_t{3});
    }
    const auto* second = find_symbol(sa, "Second", "class");
    CE_EXPECT_TRUE(second != nullptr);
    if (second != nullptr && second->location.has_value()) {
        CE_EXPECT_EQ(second->location->start.line, uint32_t{6});
    }
}

/// Duplicate declarations + forward declarations: each entity yields at most
/// one emitted symbol. Prefer the definition when one exists in the main
/// file; otherwise the earliest declaration in the main file.
void test_forward_and_duplicate_declarations() {
    ce_cpp::CompileProfile profile = build_test_profile();
    auto la = run_target(profile, "libraries/cpp/symbols/forward.hpp");
    const auto& sa = la.symbol_analysis;
    // `class NotDefined;` — no definition, single forward decl emitted.
    size_t not_def_count = 0;
    for (const auto& s : sa.symbols) {
        if (s.name == "NotDefined") ++not_def_count;
    }
    CE_EXPECT_EQ(not_def_count, size_t{1});
    // `class Defined;` + `class Defined { … };` — one symbol at the
    // definition line.
    size_t defined_count = 0;
    uint32_t defined_line = 0;
    for (const auto& s : sa.symbols) {
        if (s.name == "Defined" && s.kind == "class") {
            ++defined_count;
            if (s.location.has_value()) defined_line = s.location->start.line;
        }
    }
    CE_EXPECT_EQ(defined_count, size_t{1});
    CE_EXPECT_TRUE(defined_line >= 5);
    // Three redeclarations of `plain` collapse to one emit at the definition.
    size_t plain_count = 0;
    uint32_t plain_line = 0;
    for (const auto& s : sa.symbols) {
        if (s.name == "plain" && s.kind == "function") {
            ++plain_count;
            if (s.location.has_value()) plain_line = s.location->start.line;
        }
    }
    CE_EXPECT_EQ(plain_count, size_t{1});
    CE_EXPECT_TRUE(plain_line >= 10);
}

/// Symbols from included headers must not leak into the target's analysis.
/// The target only owns declarations physically inside its own file.
void test_included_header_symbols_not_leaked() {
    ce_cpp::CompileProfile profile = build_test_profile();
    auto la = run_target(profile, "libraries/cpp/symbols/headers.hpp");
    const auto& sa = la.symbol_analysis;
    // Local decls appear.
    CE_EXPECT_TRUE(find_symbol(sa, "LocalOnly", "class") != nullptr);
    CE_EXPECT_TRUE(find_symbol(sa, "local_fn", "function") != nullptr);
    // Included basic.hpp decls must not appear.
    CE_EXPECT_TRUE(find_symbol(sa, "algebra", "type") == nullptr);
    CE_EXPECT_TRUE(find_symbol(sa, "Point", "class") == nullptr);
    CE_EXPECT_TRUE(find_symbol(sa, "Addable", "concept") == nullptr);
}

/// Parse recovery: an error mid-file leaves earlier decls intact and marks
/// the analysis `partial` so the caller can see the catalog is incomplete.
/// Dependency completeness is not tested here — that lives in the usecases
/// integration test — but the state contract must hold.
void test_parse_recovery_is_partial() {
    ce_cpp::CompileProfile profile = build_test_profile();
    auto la = run_target(profile, "libraries/cpp/symbols/recovery.hpp");
    const auto& sa = la.symbol_analysis;
    CE_EXPECT_TRUE(find_symbol(sa, "Before", "class") != nullptr);
    CE_EXPECT_TRUE(find_symbol(sa, "before_fn", "function") != nullptr);
    // Clang's recovery is best-effort — either the analysis state is
    // `partial` (typical) or the driver bailed and we saw `failed`. Either
    // must not be `complete`.
    CE_EXPECT_TRUE(sa.state != ce_cpp::AnalysisState::Complete);
}

void run_task2_checks() {
    test_macro_expansions_are_filtered();
    test_crlf_line_numbers();
    test_forward_and_duplicate_declarations();
    test_included_header_symbols_not_leaked();
    test_parse_recovery_is_partial();
}

void run_test() {
    ce_cpp::CompileProfile profile = build_test_profile();

    std::vector<std::string> targets = {
        "libraries/cpp/symbols/basic.hpp",
        "libraries/cpp/symbols/nested.hpp",
        "libraries/cpp/symbols/unicode.hpp",
    };

    std::vector<ce_cpp::LibraryAnalysis> library_analyses;
    for (const auto& t : targets) {
        library_analyses.push_back(run_target(profile, t));
    }

    run_semantic_checks(library_analyses);
    run_task2_checks();

    std::string actual = serialize_payload(library_analyses);

    if (std::getenv("UPDATE_EXPECT") != nullptr) {
        write_file(expected_fixture_path(), actual + "\n");
        std::cerr << "wrote fixture to " << expected_fixture_path() << std::endl;
        return;
    }

    std::string expected = chomp(read_file(expected_fixture_path()));

    if (actual != expected) {
        std::cerr << "response fixture drift.\n\n--- expected ---\n"
                  << expected << "\n\n--- actual ---\n"
                  << actual << "\n";
        ++failures;
    }
}

}  // namespace

int main() {
    run_test();
    if (failures > 0) {
        std::cerr << failures << " C++ symbols assertions failed" << std::endl;
        return EXIT_FAILURE;
    }
    return EXIT_SUCCESS;
}

File: tools/library-analyzers/rust/tests/symbols.rs
//! Fixture-driven integration tests for the Rust symbol analyzer (plan 044).
//!
//! `fixture_matches_checked_in_response` loads `rust-symbols-request.json`,
//! substitutes the fixture tree's absolute path into `repository_root`, runs
//! `analyze_request`, and diffs the assembled response against
//! `rust-symbols-response.json` via `serde_json::Value` equality. Set
//! `UPDATE_EXPECT=1` to rewrite the fixture from a green run.

use std::fs;
use std::path::{Path, PathBuf};

use ce_library_rust_analyzer::dependencies::analyze_request;
use ce_library_rust_analyzer::module_graph::RustWorkspace;
use ce_library_rust_analyzer::symbols::analyze_symbols;
use library_adapter_protocol::{AnalysisRequest, AnalysisState, LibraryAnalysis, SolutionAnalysis};

fn fixture_root() -> PathBuf {
    Path::new(env!("CARGO_MANIFEST_DIR")).join("tests/fixtures")
}

fn tree_root() -> PathBuf {
    fixture_root().join("tree")
}

fn repo_root() -> PathBuf {
    let mut p = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    for _ in 0..3 {
        assert!(p.pop(), "unable to locate repo root");
    }
    p
}

fn load_request() -> AnalysisRequest {
    let protocol_fixture =
        repo_root().join("tools/library-analyzers/protocol/fixtures/rust-symbols-request.json");
    let raw = fs::read_to_string(&protocol_fixture)
        .unwrap_or_else(|e| panic!("read {}: {e}", protocol_fixture.display()));
    let substituted = raw.replace("<REPOSITORY_ROOT>", &tree_root().display().to_string());
    serde_json::from_str(&substituted).expect("request JSON parses")
}

fn expected_fixture_path() -> PathBuf {
    repo_root().join("tools/library-analyzers/protocol/fixtures/rust-symbols-response.json")
}

fn load_expected() -> Option<serde_json::Value> {
    let path = expected_fixture_path();
    let raw = fs::read_to_string(&path).ok()?;
    Some(serde_json::from_str(&raw).expect("expected response JSON parses"))
}

#[derive(serde::Serialize)]
struct ResolvedResponse {
    libraries: Vec<LibraryAnalysis>,
    solutions: Vec<SolutionAnalysis>,
}

fn resolve(request: &AnalysisRequest) -> ResolvedResponse {
    let workspace = RustWorkspace::from_request(request).expect("workspace builds");
    let (libraries, solutions) = analyze_request(request, &workspace);
    ResolvedResponse {
        libraries,
        solutions,
    }
}

#[test]
fn fixture_matches_checked_in_response() {
    let request = load_request();
    let actual = resolve(&request);
    let actual_value = serde_json::to_value(&actual).expect("serialize");

    if std::env::var_os("UPDATE_EXPECT").is_some() {
        let pretty = serde_json::to_string_pretty(&actual_value).unwrap() + "\n";
        fs::write(expected_fixture_path(), pretty).expect("write updated fixture");
        return;
    }

    let expected = load_expected().unwrap_or_else(|| {
        panic!(
            "expected fixture missing at {}. Run with UPDATE_EXPECT=1 to create it.",
            expected_fixture_path().display()
        )
    });

    if actual_value != expected {
        let actual_pretty = serde_json::to_string_pretty(&actual_value).unwrap();
        let expected_pretty = serde_json::to_string_pretty(&expected).unwrap();
        panic!(
            "response fixture drift.\n\n--- expected ---\n{expected_pretty}\n\n--- actual ---\n{actual_pretty}\n"
        );
    }
}

// ─── Focused unit tests ────────────────────────────────────────────────────

fn symbol_names(source: &str) -> Vec<(String, String)> {
    let analysis = analyze_symbols(source, "test.rs", &[]);
    analysis
        .symbols
        .into_iter()
        .map(|s| (s.name, s.kind))
        .collect()
}

#[test]
fn struct_enum_union_are_extracted() {
    let src = "pub struct S; pub enum E { A } pub union U { a: u32 }";
    let names = symbol_names(src);
    assert!(names.contains(&("S".into(), "struct".into())));
    assert!(names.contains(&("E".into(), "enum".into())));
    assert!(names.contains(&("A".into(), "enum_variant".into())));
    assert!(names.contains(&("U".into(), "union".into())));
}

#[test]
fn trait_items_get_kind_projections() {
    let src = "pub trait T { fn m(); type X; const N: u32; }";
    let names = symbol_names(src);
    assert!(names.contains(&("T".into(), "trait".into())));
    assert!(names.contains(&("m".into(), "method".into())));
    assert!(names.contains(&("X".into(), "type".into())));
    assert!(names.contains(&("N".into(), "const".into())));
}

#[test]
fn impl_block_is_not_emitted_but_methods_are() {
    let src = "pub struct S; impl S { pub fn m(&self) {} } impl Clone for S { fn clone(&self) -> Self { S } }";
    let analysis = analyze_symbols(src, "t.rs", &[]);
    let kinds: Vec<&str> = analysis.symbols.iter().map(|s| s.kind.as_str()).collect();
    assert!(
        !kinds.contains(&"impl"),
        "impl blocks must not become symbols"
    );
    // Two `method` entries: the inherent `m` and the trait-impl `clone`.
    let method_count = kinds.iter().filter(|k| **k == "method").count();
    assert_eq!(method_count, 2, "got kinds {kinds:?}");
}

#[test]
fn impl_methods_qualify_under_target_type() {
    let src = "pub struct S; impl S { pub fn m() {} }";
    let analysis = analyze_symbols(src, "t.rs", &["algebra".into()]);
    let m = analysis
        .symbols
        .iter()
        .find(|s| s.name == "m")
        .expect("method m");
    assert_eq!(m.qualified_name.as_deref(), Some("algebra::S::m"));
    assert!(m.search_names.contains(&"m".to_string()));
    assert!(m.search_names.contains(&"algebra::S::m".to_string()));
}

#[test]
fn nested_mod_extends_qualified_name() {
    let src = "pub mod outer { pub mod inner { pub fn deep() {} } }";
    let analysis = analyze_symbols(src, "t.rs", &[]);
    let deep = analysis
        .symbols
        .iter()
        .find(|s| s.name == "deep")
        .expect("nested fn");
    assert_eq!(deep.qualified_name.as_deref(), Some("outer::inner::deep"));
}

#[test]
fn macro_declaration_emits_but_macro_invocation_marks_partial() {
    // `macro_rules! shout { ... }` is a declaration → emit macro symbol.
    let src = "macro_rules! shout { () => {} }";
    let analysis = analyze_symbols(src, "t.rs", &[]);
    assert!(
        analysis
            .symbols
            .iter()
            .any(|s| s.name == "shout" && s.kind == "macro"),
    );
    assert!(matches!(analysis.state, AnalysisState::Complete));

    // Item-level `foo!()` invocation may hide further items — partial.
    let src = "lazy_static::x! {}\npub struct Public;";
    let analysis = analyze_symbols(src, "t.rs", &[]);
    assert!(matches!(analysis.state, AnalysisState::Partial));
    assert!(analysis.symbols.iter().any(|s| s.name == "Public"));
}

#[test]
fn unicode_columns_are_unicode_scalar_values() {
    // The `pub struct` item starts at column 1 — verify the walker uses
    // `ItemStruct::span()` (item start) rather than the identifier position,
    // and that the emitted column is 1-based even when the identifier itself
    // is multi-byte Unicode.
    let src = "pub struct 東京 { pub 人口: u64 }";
    let analysis = analyze_symbols(src, "t.rs", &[]);
    let tokyo = analysis
        .symbols
        .iter()
        .find(|s| s.name == "東京")
        .expect("struct 東京");
    let loc = tokyo.location.as_ref().expect("has location");
    assert_eq!(loc.start.line, 1);
    assert_eq!(loc.start.column, Some(1));
}

// ─── Task 2: validate locations and partial states ─────────────────────────

#[test]
fn line_and_column_are_one_based_at_top_of_file() {
    let src = "pub struct Head;\npub fn body() {}\n";
    let analysis = analyze_symbols(src, "t.rs", &[]);
    let head = analysis
        .symbols
        .iter()
        .find(|s| s.name == "Head")
        .expect("struct Head");
    let loc = head.location.as_ref().expect("has location");
    assert_eq!(loc.start.line, 1, "first line is 1-based");
    assert_eq!(loc.start.column, Some(1), "first column is 1-based");

    let body = analysis
        .symbols
        .iter()
        .find(|s| s.name == "body")
        .expect("fn body");
    let loc = body.location.as_ref().expect("has location");
    assert_eq!(loc.start.line, 2);
    assert_eq!(loc.start.column, Some(1));
}

#[test]
fn crlf_line_endings_count_as_one_line_each() {
    // CRLF between items — proc-macro2 treats "\r\n" as one line ending, so
    // the second item should still show up on line 2, not line 3.
    let src = "pub struct A;\r\npub struct B;\r\n";
    let analysis = analyze_symbols(src, "t.rs", &[]);
    let b = analysis
        .symbols
        .iter()
        .find(|s| s.name == "B")
        .expect("struct B");
    assert_eq!(b.location.as_ref().unwrap().start.line, 2);
}

#[test]
fn duplicate_names_are_kept_and_differentiated_by_qualified_name() {
    let src = "mod a { pub struct X; } mod b { pub struct X; }";
    let analysis = analyze_symbols(src, "t.rs", &[]);
    let xs: Vec<&library_adapter_protocol::Symbol> =
        analysis.symbols.iter().filter(|s| s.name == "X").collect();
    assert_eq!(xs.len(), 2, "both X symbols preserved");
    let qualifieds: Vec<&str> = xs
        .iter()
        .filter_map(|s| s.qualified_name.as_deref())
        .collect();
    assert!(qualifieds.contains(&"a::X"));
    assert!(qualifieds.contains(&"b::X"));
}

#[test]
fn malformed_source_yields_failed_and_no_symbols() {
    // Missing closing brace — syn's parser cannot recover.
    let src = "pub struct Broken {";
    let analysis = analyze_symbols(src, "t.rs", &[]);
    assert!(matches!(analysis.state, AnalysisState::Failed));
    assert!(analysis.symbols.is_empty());
}

#[test]
fn public_and_private_visibilities_are_both_extracted() {
    // The adapter does not filter by visibility — the core layer decides
    // what to publish. Verify both come through with kind + name intact.
    let src = "pub struct P; struct Q; pub(crate) fn r() {} fn s() {}";
    let analysis = analyze_symbols(src, "t.rs", &[]);
    let names: Vec<&str> = analysis.symbols.iter().map(|s| s.name.as_str()).collect();
    for expected in ["P", "Q", "r", "s"] {
        assert!(names.contains(&expected), "missing {expected} in {names:?}");
    }
}

#[test]
fn file_backed_mod_declarations_do_not_leak_other_files_symbols() {
    // `mod other;` without content — the analyzer must not open `other.rs`.
    // Only the `mod` declaration itself is emitted; children stay silent.
    let src = "mod other;\npub struct Local;\n";
    let analysis = analyze_symbols(src, "t.rs", &[]);
    let names: Vec<&str> = analysis.symbols.iter().map(|s| s.name.as_str()).collect();
    assert!(names.contains(&"other"), "file-backed mod is emitted");
    assert!(names.contains(&"Local"), "sibling items still appear");
    // No off-file symbol from a hypothetical `other.rs` may appear.
    assert_eq!(analysis.symbols.len(), 2);
}

#[test]
fn locations_are_never_reversed() {
    // Every emitted location must have end >= start.
    let src = include_str!("fixtures/tree/libraries/rust/symbols/basic.rs");
    let analysis = analyze_symbols(src, "basic.rs", &[]);
    for symbol in &analysis.symbols {
        let loc = symbol
            .location
            .as_ref()
            .expect("all basic.rs symbols located");
        let end = loc.end.as_ref().expect("basic.rs spans have end");
        assert!(
            (end.line, end.column) >= (loc.start.line, loc.start.column),
            "reversed span for {:?}",
            symbol.name
        );
    }
}

// ─── Issue #105: analyze_request wire-up ───────────────────────────────────

fn write_library_tree(files: &[(&str, &str)]) -> tempfile::TempDir {
    let dir = tempfile::tempdir().expect("tempdir");
    for (rel, body) in files {
        let abs = dir.path().join(rel);
        std::fs::create_dir_all(abs.parent().unwrap()).expect("mkdir -p");
        std::fs::write(&abs, body).expect("write library file");
    }
    dir
}

fn request_with_library(repo_root: &std::path::Path, library_path: &str) -> AnalysisRequest {
    AnalysisRequest {
        schema_version: library_adapter_protocol::SCHEMA_VERSION,
        language: "rust".into(),
        repository_root: repo_root.display().to_string(),
        libraries: vec![library_adapter_protocol::LibraryTarget {
            path: library_path.into(),
        }],
        solutions: vec![],
    }
}

#[test]
fn analyze_request_emits_partial_with_diagnostic_on_item_level_macro() {
    // `lazy_static::x!{}` is an item-level macro invocation the syn walker
    // cannot expand. The walker marks the analysis Partial and the wire-up
    // must surface a diagnostic so downstream UI can flag the gap.
    let tree = write_library_tree(&[(
        "libraries/rust/partial.rs",
        "lazy_static::x!{}\npub struct Kept;\n",
    )]);
    let request = request_with_library(tree.path(), "libraries/rust/partial.rs");
    let workspace = RustWorkspace::from_request(&request).expect("workspace builds");
    let (libraries, _solutions) = analyze_request(&request, &workspace);

    let lib = libraries.first().expect("one library analyzed");
    assert!(
        matches!(lib.symbol_analysis.state, AnalysisState::Partial),
        "state was {:?}",
        lib.symbol_analysis.state
    );
    assert!(
        lib.symbol_analysis.symbols.iter().any(|s| s.name == "Kept"),
        "walker still emits items it could see: {:?}",
        lib.symbol_analysis.symbols,
    );
    let symbol_diags: Vec<&library_adapter_protocol::Diagnostic> = lib
        .diagnostics
        .iter()
        .filter(|d| d.code == "rust.symbols.partial")
        .collect();
    assert_eq!(
        symbol_diags.len(),
        1,
        "exactly one rust.symbols.partial diagnostic expected, got {:?}",
        lib.diagnostics,
    );
    assert!(matches!(
        symbol_diags[0].severity,
        library_adapter_protocol::Severity::Warning
    ));
}

#[test]
fn analyze_request_emits_failed_with_diagnostic_on_broken_source_without_cascading_into_dependencies()
 {
    // An unterminated struct body trips `syn::parse_file`. Both the
    // dependency pass and the symbol pass see the same failure, but each
    // reports its own diagnostic — proving the two pipelines are wired
    // independently rather than sharing one failure state.
    let tree = write_library_tree(&[(
        "libraries/rust/broken.rs",
        "pub struct Broken {\n",
    )]);
    let request = request_with_library(tree.path(), "libraries/rust/broken.rs");
    let workspace = RustWorkspace::from_request(&request).expect("workspace builds");
    let (libraries, _solutions) = analyze_request(&request, &workspace);

    let lib = libraries.first().expect("one library analyzed");
    assert!(
        matches!(lib.symbol_analysis.state, AnalysisState::Failed),
        "state was {:?}",
        lib.symbol_analysis.state
    );
    assert!(lib.symbol_analysis.symbols.is_empty());

    let symbol_diag = lib
        .diagnostics
        .iter()
        .find(|d| d.code == "rust.symbols.parse")
        .unwrap_or_else(|| {
            panic!(
                "expected rust.symbols.parse diagnostic, got {:?}",
                lib.diagnostics
            )
        });
    assert!(matches!(
        symbol_diag.severity,
        library_adapter_protocol::Severity::Warning
    ));

    // The dependency pass already emits its own diagnostic on the same
    // broken file — assert both codes coexist on the same LibraryAnalysis.
    assert!(
        lib.diagnostics
            .iter()
            .any(|d| d.code == "rust.parse.entry_file"),
        "dependency pass still emits its own diagnostic: {:?}",
        lib.diagnostics,
    );
}
